java-manta-client-unshaded/src/main/java/com/joyent/manta/exception/MantaUnexpectedObjectTypeException.java

```diff
@@ -0,0 +1,160 @@
+     *
+     * @param message the detail message. The detail message is saved for
+    *                 later retrieval by the {@link #getMessage()} method.
+     * @param expected expected object type
+     * @param actual actual object type
+     */
+    public MantaUnexpectedObjectTypeException(final String message,
+                                              final ObjectType expected,
+                                              final ObjectType actual) {
+        super(message);
+        this.expected = expected;
+        this.actual = actual;
+        updateContext();
+    }
+
+    /**
+     * Creates a new instance with the expected properties.
+     *
+     * @param cause the cause (which is saved for later retrieval by the
+     *              {@link #getCause()} method).  (A <tt>null</tt> value is
+     *              permitted, and indicates that the cause is nonexistent or
+     *              unknown.)
+     * @param expected expected object type
+     * @param actual actual object type
+     */
+    public MantaUnexpectedObjectTypeException(final Throwable cause,
+                                              final ObjectType expected,
+                                              final ObjectType actual) {
+        super(msg(expected, actual), cause);
+        this.expected = expected;
+        this.actual = actual;
+        updateContext();
+    }
+
+    /**
+     * Creates a new instance with the expected properties.
+     *
+     * @param message the detail message. The detail message is saved for
+     *                later retrieval by the {@link #getMessage()} method.
+     * @param cause the cause (which is saved for later retrieval by the
+     *              {@link #getCause()} method).  (A <tt>null</tt> value is
+     *              permitted, and indicates that the cause is nonexistent or
+     *              unknown.)
+     * @param expected expected object type
+     * @param actual actual object type
+     */
+    public MantaUnexpectedObjectTypeException(final String message,
+                                              final Throwable cause,
+                                              final ObjectType expected,
+                                              final ObjectType actual) {
+        super(message, cause);
+        this.expected = expected;
+        this.actual = actual;
+        updateContext();
+    }
+
+    public ObjectType getExpected() {
+        return expected;
+    }
+
+    public ObjectType getActual() {
+        return actual;
+    }
+
+    public MantaHttpHeaders getResponseHeaders() {
+        return responseHeaders;
+    }
+
+    /**
+     * Sets the response headers of the failed request. This is useful for
+     * debugging the reason of why the objects didn't match the expectation.
+     *
+     * @param responseHeaders headers object to attach to exception
+     */
+    public void setResponseHeaders(final MantaHttpHeaders responseHeaders) {
```

Comment:
This method is really nice because it allows consumers to salvage information about the request out of the failure mode.